File: tests/skipping-stones.test.js
import test from 'node:test';
import assert from 'node:assert/strict';

import {
  SKIP_RULES_VERSION,
  SKIP_CLEANUP_TICKS,
  SKIP_LAUNCH_BOUNDS,
  DEFAULT_SKIP_ENVIRONMENT,
  initSkippingState,
  validateSkippingInput,
  validateSkipLaunch,
  simulateSkipTrajectory,
  applySkippingInput,
  clearThrow,
  stepSkippingSimulation,
} from '../shared/skippingStonesModel.js';

const CALM = { ...DEFAULT_SKIP_ENVIRONMENT, wind: [0, 0], windSpeed: 0, rain: 0, frozenAt: 10 };
const WINDY = { ...DEFAULT_SKIP_ENVIRONMENT, wind: [0.8, -0.2], windSpeed: 0.82, rain: 0, frozenAt: 10 };
const RAINY = { ...DEFAULT_SKIP_ENVIRONMENT, wind: [0, 0], windSpeed: 0, rain: 0.95, frozenAt: 10 };

const LAUNCH = { kind: 'launch', angle: 0.2, power: 0.85, yaw: 0 };

test('skipping stones: validate clamps angle/power and rejects junk', () => {
  assert.equal(validateSkipLaunch(null).valid, false);
  assert.equal(validateSkippingInput({ kind: 'teleport' }).valid, false);

  const clamped = validateSkipLaunch({ angle: 4, power: -1, yaw: 2 });
  assert.equal(clamped.valid, true);
  assert.equal(clamped.sanitized.angle, SKIP_LAUNCH_BOUNDS.maxAngle);
  assert.equal(clamped.sanitized.power, SKIP_LAUNCH_BOUNDS.minPower);
  assert.equal(clamped.sanitized.yaw, 0.35);
});

test('skipping stones: identical launches under frozen wind/rain reproduce skip count and distance', () => {
  const origin = [0, 0.95, 0.15];
  const a = simulateSkipTrajectory(LAUNCH, CALM, origin);
  const b = simulateSkipTrajectory(LAUNCH, CALM, origin);
  assert.equal(a.skips, b.skips);
  assert.equal(a.distance, b.distance);
  assert.deepEqual(a.landingPos, b.landingPos);
  assert.equal(a.trajectory.length, b.trajectory.length);
  assert.deepEqual(a.trajectory[12], b.trajectory[12]);
  assert.ok(a.skips >= 1, `calm water should skip at least once, got ${a.skips}`);
});

test('skipping stones: frozen wind and rain change shared outcomes deterministically', () => {
  const origin = [0, 0.95, 0.15];
  const calm = simulateSkipTrajectory(LAUNCH, CALM, origin);
  const wind = simulateSkipTrajectory(LAUNCH, WINDY, origin);
  const rain = simulateSkipTrajectory(LAUNCH, RAINY, origin);

  assert.notEqual(calm.distance, wind.distance);
  assert.ok(wind.landingPos[0] > calm.landingPos[0], 'east wind should push the stone +X');
  assert.ok(
    rain.skips < calm.skips || rain.distance < calm.distance,
    `rain should shorten the run: calm ${calm.skips}/${calm.distance} rain ${rain.skips}/${rain.distance}`,
  );
});

test('skipping stones: two throwers share outcomes and leave independently', () => {
  let state = initSkippingState({ activeSlots: [0, 1], environment: CALM, nowMs: 10 });
  assert.equal(state.rulesVersion, SKIP_RULES_VERSION);
  assert.equal(state.environment.policy, 'frozen');

  const first = applySkippingInput(state, 0, LAUNCH);
  state = first.simState;
  assert.equal(first.event?.type, 'stone_launched');
  assert.equal(state.throwers[0].currentThrow.skips, first.event.payload.skips);

  const second = applySkippingInput(state, 1, { ...LAUNCH, power: 0.7 });
  state = second.simState;
  assert.ok(state.throwers[0].currentThrow);
  assert.ok(state.throwers[1].currentThrow);
  assert.equal(state.throwers[0].currentThrow.distance, first.event.payload.distance);

  state = clearThrow(state, 1);
  assert.equal(state.throwers[1].currentThrow, null);
  assert.equal(state.throwers[1].phase, 'idle');
  assert.ok(state.throwers[0].currentThrow);
});

test('skipping stones: live weather after start cannot rewrite frozen conditions', () => {
  let state = initSkippingState({ activeSlots: [0], environment: CALM, nowMs: 10 });
  const before = state.environment;
  state = applySkippingInput(state, 0, {
    kind: 'environment',
    environment: WINDY,
  }).simState;
  assert.deepEqual(state.environment, before);

  const replay = simulateSkipTrajectory(LAUNCH, state.environment, state.throwers[0].origin);
  const expected = simulateSkipTrajectory(LAUNCH, CALM, state.throwers[0].origin);
  assert.equal(replay.skips, expected.skips);
  assert.equal(replay.distance, expected.distance);
});

test('skipping stones: sunk throws clean up after the shared timeout', () => {
  let state = initSkippingState({ activeSlots: [0], environment: CALM });
  state = applySkippingInput(state, 0, LAUNCH).simState;
  const flightMs = state.throwers[0].currentThrow.flightTimeMs;
  const flightSteps = Math.ceil(flightMs / (1000 / 60)) + 1;

  let res = stepSkippingSimulation(state, {}, flightSteps);
  state = res.simState;
  assert.ok(res.events.some((e) => e.type === 'stone_sunk'));
  assert.equal(state.throwers[0].phase, 'sunk');
  assert.ok(state.throwers[0].lastThrow.distance > 0);

  res = stepSkippingSimulation(state, {}, SKIP_CLEANUP_TICKS);
  state = res.simState;
  assert.ok(res.events.some((e) => e.type === 'stone_cleared'));
  assert.equal(state.throwers[0].currentThrow, null);
  assert.equal(state.throwers[0].phase, 'idle');
});

File: server_elixir/test/afterlight/activities/skipping_stones_test.exs
defmodule Afterlight.Activities.SkippingStonesTest do
  use ExUnit.Case, async: true

  alias Afterlight.Activities.SkippingStones

  @calm %{"wind" => [0.0, 0.0], "windSpeed" => 0.0, "rain" => 0.0, "frozenAt" => 10}
  @windy %{"wind" => [0.8, -0.2], "windSpeed" => 0.82, "rain" => 0.0, "frozenAt" => 10}
  @rainy %{"wind" => [0.0, 0.0], "windSpeed" => 0.0, "rain" => 0.95, "frozenAt" => 10}
  @launch %{"kind" => "launch", "angle" => 0.2, "power" => 0.85, "yaw" => 0.0}
  @origin [0.0, 0.95, 0.15]

  describe "init_sim_state/1 and validate_controls/1" do
    test "freezes start conditions and clamps launch bounds" do
      state = SkippingStones.init_sim_state(slots: [0, 1], environment: @calm, now_ms: 10)
      assert state["status"] == "open"
      assert state["environment"]["policy"] == "frozen"
      assert state["environment"]["frozenAt"] == 10

      assert {:ok, clamped} =
               SkippingStones.validate_controls(%{"kind" => "launch", "angle" => 4, "power" => -1, "yaw" => 2})

      assert clamped["angle"] == 0.55
      assert clamped["power"] == 0.15
      assert clamped["yaw"] == 0.35
      assert {:error, :unknown_kind} = SkippingStones.validate_controls(%{"kind" => "teleport"})
    end
  end

  describe "simulate_skip/3" do
    test "identical launches under frozen wind/rain reproduce skip count and distance" do
      a = SkippingStones.simulate_skip(@launch, @calm, @origin)
      b = SkippingStones.simulate_skip(@launch, @calm, @origin)
      assert a.skips == b.skips
      assert a.distance == b.distance
      assert a.landing_pos == b.landing_pos
      assert length(a.trajectory) == length(b.trajectory)
      assert a.skips >= 1
    end

    test "frozen wind and rain change outcomes deterministically" do
      calm = SkippingStones.simulate_skip(@launch, @calm, @origin)
      wind = SkippingStones.simulate_skip(@launch, @windy, @origin)
      rain = SkippingStones.simulate_skip(@launch, @rainy, @origin)

      assert calm.distance != wind.distance
      [calm_x, _, _] = calm.landing_pos
      [wind_x, _, _] = wind.landing_pos
      assert wind_x > calm_x
      assert rain.skips < calm.skips or rain.distance < calm.distance
    end
  end

  describe "apply_input/3 and step_simulation/3" do
    test "two throwers share outcomes; leave and cleanup are independent" do
      state = SkippingStones.init_sim_state(slots: [0, 1], environment: @calm, now_ms: 10)
      {:ok, state} = SkippingStones.apply_input(state, 0, @launch)
      assert state["throwers"]["0"]["currentThrow"]["skips"] >= 1

      {:ok, state} = SkippingStones.apply_input(state, 1, Map.put(@launch, "power", 0.7))
      assert state["throwers"]["0"]["currentThrow"]
      assert state["throwers"]["1"]["currentThrow"]

      {:ok, state} = SkippingStones.apply_input(state, 1, %{"kind" => "leave"})
      assert state["throwers"]["1"]["currentThrow"] == nil
      assert state["throwers"]["1"]["phase"] == "idle"
      assert state["throwers"]["0"]["currentThrow"]
    end

    test "later weather cannot rewrite frozen start conditions" do
      state = SkippingStones.init_sim_state(slots: [0], environment: @calm, now_ms: 10)
      before = state["environment"]
      {:ok, state} = SkippingStones.apply_input(state, 0, %{"kind" => "environment", "environment" => @windy})
      assert state["environment"] == before

      origin = state["throwers"]["0"]["origin"]
      replay = SkippingStones.simulate_skip(@launch, state["environment"], origin)
      expected = SkippingStones.simulate_skip(@launch, @calm, origin)
      assert replay.skips == expected.skips
      assert replay.distance == expected.distance
    end

    test "sunk throws clean up after the shared timeout" do
      state = SkippingStones.init_sim_state(slots: [0], environment: @calm)
      {:ok, state} = SkippingStones.apply_input(state, 0, @launch)
      flight_ms = state["throwers"]["0"]["currentThrow"]["flightTimeMs"]
      steps = ceil(flight_ms / (1000 / 60)) + 4

      {state, outcome} = SkippingStones.step_simulation(state, %{}, steps)
      assert outcome == nil
      assert state["throwers"]["0"]["phase"] == "sunk"
      assert state["throwers"]["0"]["lastThrow"]["distance"] > 0

      {state, _} = SkippingStones.step_simulation(state, %{}, SkippingStones.cleanup_ticks())
      assert state["throwers"]["0"]["currentThrow"] == nil
      assert state["throwers"]["0"]["phase"] == "idle"
    end
  end
end
